File: proyectoRuby/lib/casilla.rb
# encoding:utf-8
require_relative 'tipo_casilla'

module Civitas
class Casilla
    
  def self.new_DESCANSO(_nombre)
    @tipo = TipoCasilla::DESCANSO
    new(nil, _nombre, nil, nil, nil)
  end
  
  def self.new_CALLE(titulo)
    @tipo = TipoCasilla::CALLE
    new(nil, titulo.nombre, titulo, nil, nil)
  end
  
  def self.new_IMPUESTO(cantidad, nombre)
    @tipo = TipoCasilla::IMPUESTO
    new(nil, nombre, nil, cantidad, nil)
  end
  
  def self.new_JUEZ(numCasillaCarcel, nombre)
    @tipo = TipoCasilla::JUEZ
    new(nil, nombre, nil, nil, numCasillaCarcel)
  end
  
  def self.new_SORPRESA(mazo, nombre)
    @tipo = TipoCasilla::SORPRESA
    new(mazo, nombre, nil, nil, nil)
  end
  
  # Constructor
  def initialize(mazo, nombre, titulo, cantidad, numCasillaCarcel)
    init
    @nombre = nombre
    @sorpresa = nil #tipo sorpresa
    @mazo = mazo  # tipo sorpresa
    @tituloPropiedad = titulo #tipo calle
    @@carcel = numCasillaCarcel #tipo juez
    @importe = cantidad
  end
  
  private :initialize
  
  # Métodos
  def informe(actual, todos)
    if jugador_correcto(actual, todos)
      nombre_jugador = todos[actual].nombre
      evento = "El jugador " + nombre_jugador + " ha caído en la casilla " + to_string
      Diario.instance.ocurre_evento(evento)
    end
  end
  
  def init
    @importe = 0
    @sorpresa = nil
    @mazo = nil
    @tituloPropiedad = nil
    @tipo = nil
    @@carcel = 0
    @nombre = "casilla"
  end
  
  def jugador_correcto(actual, todos)
    if actual < todos.length && actual >= 0
      return true
    else
      return false
    end
  end
  
  def recibe_jugador(actual, todos)
    #no implementado
  end
  
  def recibe_jugador_calle(actual, todos)
    # no implementado
  end
  
  def recibe_jugador_impuesto(actual, todos)
    if jugador_correcto(actual, todos)
      informe(actual, todos)
      todos[actual].paga_impuesto(@importe)
    end
  end
  
  def recibe_jugador_juez(actual, todos)
    if jugador_correcto(actual, todos)
      informe(actual, todos)
      todos[actual].encarcelar(@@carcel)
    end
  end
  
  def recibe_jugador_sorpresa(actual, todos)
    # no implementado
  end
  
  def to_string
    return @nombre
  end
  
  # Consultor
  attr_reader :nombre, :tituloPropieddad
end
end

File: proyectoJava/src/civitas/Casilla.java
package civitas;

import java.util.ArrayList;

public class Casilla {
    
    // Atributos
    static private int carcel;
    private float importe;
    private String nombre;
    TipoCasilla tipo;
    
    MazoSorpresas mazo;                 // tipo sorpresa
    Sorpresa sorpresa;                  // tipo sorpresa
    TituloPropiedad tituloPropiedad;    // tipo calle
    
    /**
     * Constructor
     * @param s nombre de la casilla 
     */
    Casilla ( String s ){   //descanso
        init();
        nombre = s;
        tipo = TipoCasilla.DESCANSO;
    }
    
    Casilla (TituloPropiedad titulo){   // calle
        init();
        nombre = titulo.getNombre();
        tipo = TipoCasilla.CALLE;
        tituloPropiedad = titulo;
    }
    
    Casilla (float cantidad, String nombre){    // impuesto
        init();
        tipo = TipoCasilla.IMPUESTO;
        importe = cantidad;
        nombre = "Impuesto";
    }
    
    Casilla (int numCasillaCarcel, String nombre){   // juez
        init();
        tipo = TipoCasilla.JUEZ;
        carcel = numCasillaCarcel;
        nombre = "Juez";
    }
    
    Casilla (MazoSorpresas _mazo, String _nombre){    // sorpresa
        init();
        tipo = TipoCasilla.SORPRESA;
        mazo = _mazo;
        nombre = _nombre;
        nombre = "Sorpresa";
    }
    
    public String getNombre(){
        return nombre;
    }
    
    TituloPropiedad getTituloPropiedad(){
        return tituloPropiedad;
    }
    
    private void informe (int iactual, ArrayList<Jugador> todos){
        if (jugadorCorrecto(iactual, todos)){
            String nombreJugador = todos.get(iactual).getNombre();
            String evento = "El jugador " + nombreJugador + " ha caído en la casilla " + nombre;
            Diario.getInstance().ocurreEvento(evento);
        }
    }
    
    private void init (){
        importe = 0;
        nombre = "Casilla sin nombre";
        tipo = TipoCasilla.DESCANSO;
        mazo = new MazoSorpresas();
        tituloPropiedad = new TituloPropiedad("Propiedad sin inicializar", 0, 0, 0, 0, 0);
    }
    
    /**
     * Comprueba si el primer parámetro es un índice válido para acceder al ArrayList de jugadores
     * @param iactual índice del jugador actual
     * @param todos ArrayList de jugadores
     * @return true si es un índice válido
     */
    public boolean jugadorCorrecto(int iactual, ArrayList<Jugador> todos){
        if (iactual < todos.size() && iactual >= 0)
            return true;
        else
            return false;
    }
    
    void recibeJugador (int iactual, ArrayList<Jugador> todos){
        throw new UnsupportedOperationException("No implementado");
    }
 
    private void recibeJugador_calle (int iactual, ArrayList<Jugador> todos){
        throw new UnsupportedOperationException("No implementado");
    }
    
    private void recibeJugador_impuesto (int iactual, ArrayList<Jugador> todos){
        if (jugadorCorrecto(iactual,todos)){
            informe(iactual,todos);
            todos.get(iactual).pagaImpuesto(importe);
        }
    }
    
    private void recibeJugador_juez (int iactual, ArrayList<Jugador> todos){
        if (jugadorCorrecto(iactual,todos)){
            informe(iactual,todos);
            todos.get(iactual).encarcelar(carcel);
        }
    }
    
    private void recibeJugador_sorpresa (int iactual, ArrayList<Jugador> todos){
        throw new UnsupportedOperationException("No implementado");
    }
    
    @Override
    public String toString(){
        return nombre;
    }
}
